File: Data Structures & Algorithms/count-number-of-islands/submission-0.java
class Solution {
    public int numIslands(char[][] grid) {
        int count = 0; 
        int rows = grid.length; 
        int cols = grid[0].length; 
        for(int row = 0; row < rows; row++){
            for(int col = 0; col < cols; col++){
                if(grid[row][col] == '1'){
                    dfs(grid, row, col, rows, cols);

                    count++;
                }
                else{
                    continue;
                }
            }
        }
        return count;
    }

    private void dfs(char[][] grid, int row, int col, int rows, int cols){
        if(row < 0 || row >= rows){
            return;
        }
        if(col < 0 || col >= cols){
            return;
        }
        if(grid[row][col] == '0'){
            return;
        }
        else{
            grid[row][col] = '0';

            dfs(grid, row + 1, col, rows, cols);
            dfs(grid, row - 1, col, rows, cols);
            dfs(grid, row, col + 1, rows, cols);
            dfs(grid, row, col - 1, rows, cols);
        }
    }
}

File: Data Structures & Algorithms/max-area-of-island/submission-0.py


class Solution:
    def maxAreaOfIsland(self, grid: List[List[int]]) -> int:
        ROWS, COLUMNS = len(grid), len(grid[0])
        visited = set()
        cur_max = 0

        def dfs(row, column):
            # base cases
            # check if this node is a zero
            # check if this node is inbounds
            # check if this node is visited
            if (min(row, column) < 0 or row >= ROWS or column >= COLUMNS or
                    (row, column) in visited or grid[row][column] == 0):
                return 0
            
            cur_length = 1
            visited.add((row,column))
            
            directions = [(0, 1), (0, -1), (1, 0), (-1, 0)]
            for direction in directions:
                row_add, column_add = direction
                new_row, new_column = row + row_add, column + column_add
                cur_length += dfs(new_row, new_column)
            
            return cur_length     

        for i in range(ROWS):
            for j in range(COLUMNS):
                if grid[i][j] == 1 and grid[i][j] not in visited:
                    cur_max = max(cur_max, dfs(i, j))
        
        return cur_max

        